core/src/test/java/org/apache/accumulo/core/file/rfile/GenerateSplitsTest.java

```diff
@@ -0,0 +1,148 @@
+import static org.apache.accumulo.core.file.rfile.RFileTest.newKey;
+import static org.apache.accumulo.core.file.rfile.RFileTest.newValue;
+import static org.junit.Assert.assertThrows;
+import static org.junit.Assert.assertTrue;
+
+import java.io.ByteArrayOutputStream;
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.PrintStream;
+import java.util.List;
+
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+
+@SuppressFBWarnings(value = "PATH_TRAVERSAL_IN", justification = "paths provided by test")
+public class GenerateSplitsTest {
+  private static final Logger log = LoggerFactory.getLogger(GenerateSplitsTest.class);
+
+  @ClassRule
+  public static final TemporaryFolder tempFolder =
+      new TemporaryFolder(new File(System.getProperty("user.dir") + "/target"));
+
+  private static final RFileTest.TestRFile trf = new RFileTest.TestRFile(null);
+  private static String fileName;
+
+  /**
+   * Creates a test file with 84 bytes of data and 2 Locality groups.
+   */
+  @BeforeClass
+  public static void createFile() throws IOException {
+    trf.openWriter(false);
+    trf.writer.startNewLocalityGroup("lg1", newColFamByteSequence("cf1", "cf2"));
+    trf.writer.append(newKey("r1", "cf1", "cq1", "L1", 55), newValue("foo1"));
+    trf.writer.append(newKey("r2", "cf2", "cq1", "L1", 55), newValue("foo2"));
+    trf.writer.append(newKey("r3", "cf2", "cq1", "L1", 55), newValue("foo3"));
+    trf.writer.startNewLocalityGroup("lg2", newColFamByteSequence("cf3", "cf4"));
+    trf.writer.append(newKey("r4", "cf3", "cq1", "L1", 55), newValue("foo4"));
+    trf.writer.append(newKey("r5", "cf4", "cq1", "L1", 55), newValue("foo5"));
+    trf.writer.append(newKey("r6", "cf4", "cq1", "L1", 55), newValue("foo6"));
+    trf.closeWriter();
+
+    File file = tempFolder.newFile("testGenerateSplits.rf");
+    FileOutputStream fileOutputStream = new FileOutputStream(file);
+    fileOutputStream.write(trf.baos.toByteArray());
+    fileOutputStream.flush();
+    fileOutputStream.close();
+
+    fileName = file.getAbsolutePath();
+    log.info("Wrote to file {}", fileName);
+  }
+
+  @AfterClass
+  public static void cleanUp() {
+    File file = new File(fileName);
+    if (file.delete())
+      log.info("Cleaned up test file {}", fileName);
+  }
+
+  @Test
+  public void testNum() throws Exception {
+    List<String> args = List.of(fileName, "--num", "2");
+    log.info("Invoking GenerateSplits with {}", args);
+    PrintStream oldOut = System.out;
+    try (ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        PrintStream newOut = new PrintStream(baos)) {
+      System.setOut(newOut);
```

Comment:
Changing the System output stream isn't a great idea, because it can mask output, such as the error messages printed to the console from failed assert statements.

A better solution, is to design the code to be testable by having a PrintStream parameter in the constructor. The main method will pass System.out, but for testing, you can pass something else. So, main just looks like:

```java
   public static void main(String[] args) {
     new GenerateSplits(System.out).execute(args);
   }
```

Alternatively, use the `-sf` parameter you made to create a file, and then inspect the file.

Reply:
Copied from https://github.com/apache/accumulo/blob/785a3645261571f000a7adb3c7c72b07886f0587/test/src/main/java/org/apache/accumulo/test/functional/ReadWriteIT.java#L393

Reply:
ReadWriteIT sets a bad precedence. You should not follow that precedence. Instead, PrintInfo should be updated to be more testable (like I suggest above), so that it doesn't need to do that. I checked and the only other place where we mess with the static System.out is in ShellConfigTest. We also do a similar thing with STDIN in PasswordConverterTest. Both of those should probably be fixed as well, if possible. Fixing any of those is outside the scope of this PR... I just want to make sure we don't repeat the pattern when it's avoidable.

(FWIW, I can't be certain this causes the problems I've occasionally seen where I have to run `stty sane` to fix my terminal after a `mvn package` because my keystrokes aren't echo'd to the screen, but I wouldn't be surprised if it was related.)

Reply:
Replaced the STDOUT analysis with a file in each of the tests.